crates/sparrow-session/src/execution.rs

```diff
@@ -99,29 +95,51 @@ impl Execution {
 
     pub fn next_blocking(&mut self) -> error_stack::Result<Option<RecordBatch>, Error> {
         self.is_done()?;
-        Ok(self.rt.block_on(self.output.next()))
+        Ok(self.handle.block_on(self.output.next()))
     }
 
-    pub async fn collect_all(self) -> error_stack::Result<Vec<RecordBatch>, Error> {
-        // TODO: For large outputs, we likely need to drain the output while waiting for the future.
-        match self.status {
-            Status::Running(future) => future.await?,
+    pub async fn collect_all(mut self) -> error_stack::Result<Vec<RecordBatch>, Error> {
+        let mut progress = match self.status {
+            Status::Running(progress) => progress,
             Status::Failed => error_stack::bail!(Error::ExecutionFailed),
-            _ => {}
+            Status::Completed => {
+                // If the progress channel has completed without error, we know that the output channel
+                // hasn't filled up, so we can go ahead and collect the output
+                return Ok(self.output.collect().await);
+            }
         };
 
-        Ok(self.output.collect().await)
+        let mut outputs = Vec::new();
+        loop {
+            tokio::select! {
+                // Poll futures in the order listed.
+                biased;
+
+                progress = progress.next() => {
+                    if let Some(execute_response) = progress {
+                        // Propagate errors
+                        let _ = execute_response?;
+                    }
+                }
+                batch = self.output.next() => {
+                    if let Some(batch) = batch {
+                        outputs.push(batch);
+                    } else {
+                        // Output stream is empty, break out and return
```

Comment:
Not just empty. We also need to make sure that it is *done* (eg., nothing will be added). If we get here and this gets `None`, then it is reasonable to assume the senders are dropped, so we shouldn't get more batches in the future.

But consider that they were dropped due to an error. If we finish here, we may fail to report the error if the progress channel isn't done.